### PR TITLE
hse error decoder

### PR DESCRIPTION
**Pull request by jjacob512**
Signed-off-by: Jacob M. Jacob <4000419+jjacob512@users.noreply.github.com>

### CODE REVIEW

**Comment by tristan957** [comment]
Jacob could post an output of a usage of the new command?

**Comment by jjacob512** [comment]
outputs -
$hse strerror  -- -19410277913067508
Cannot allocate memory
$hse strerror -v -- -19410277913067508
lib/c0/c0_kvmultiset.c:2676: Cannot allocate memory
$hse strerror  -hv
Usage: hse strerror [options] [--] <errorcode>

Options:
  -h, --help              Print help
  -v, --verbose           Verbose output

Note that if a file name was shown in the verbose output, it is reliable only
when the input error code was generated on this build.

**Comment by tristan957** [comment]
"this build" maybe should change to "same build"

**Comment by smoyergh** [comment]
I'd propose something like "hse util strerror" to accommodate future utility functions.

**Comment by tristan957** [comment]
Would others think it useful to also put the errno value into the hse_strerror() buffer?

**Comment by jjacob512** [comment]
> Would others think it useful to also put the errno value into the hse_strerror() buffer?
> 
> ```
> lib/c0/c0_kvmultiset.c:2676: Cannot allocate memory (12)
> ```

**Comment by tristan957** [comment]
I think the errno change should be made in hse_err.c

**Comment by alexttx** [comment]
> Note that if a file name was shown in the verbose output, it is reliable only
> when the input error code was generated on this build

Suggest:   "... if a file name is shown ..."

File: cli/hse_cli.c [hunk 1]
```diff
@@ -1447,6 +1449,97 @@ cli_hse_storage(struct cli_cmd *self, struct cli *cli)
...
+    if (parse_s64(error_code_str, &s_err)) {
+        fprintf(
+            stderr,
+            "%s: unable to parse"
+            " '%s' as a signed 64-bit"
+            " scalar value\n",
+            self->cmd_path,
+            error_code_str);
+        return EX_USAGE;
+    }
```
Comment (alexttx) [inline cli/hse_cli.c:1529]: This looks formatted for 24 column output.  Suggest not splitting the string constant and eliminating un-necessary newlines.

**Review by alexttx — APPROVED** [review]

**Comment by beckerg** [comment]
Honestly, I don't think the non-verbose form of the error message is at all useful.  If you give that to me I'm just going to have to ask you to run it again with -v.  If dealing with a customer would rather get it right the first time rather than have to go around in circles to get what we need.

FWIW, both the file name and line are suspect when run against a build different than the one that generated the err code.

**Review by beckerg — CHANGES_REQUESTED** [review]

**Comment by jjacob512** [comment]
> C



> Honestly, I don't think the non-verbose form of the error message is at all useful. If you give that to me I'm just going to have to ask you to run it again with -v. If dealing with a customer would rather get it right the first time rather than have to go around in circles to get what we need.
> 
> FWIW, both the file name and line are suspect when run against a build different than the one that generated the err code.

The -v was meant to hide un-reliable information. However I can be convinced otherwise since this is a developer only tool and its going behind the utils section. RE: line number, the line number is the actual number encoded in the error code, it is in that sense accurate w.r.t the code base on which the error code was generated. The file name can become unreliable.

File: cli/hse_cli.c [hunk 2]
```diff
@@ -1447,6 +1449,97 @@ cli_hse_storage(struct cli_cmd *self, struct cli *cli)
...
+                return EX_USAGE;
+        }
+    }
+
+    error_code_str = cli_next_arg(cli);
+
+    if (!error_code_str || help) {
+        int rc = help ? 0 : EX_USAGE;
+
+        cmd_print_help(self, help ? stdout : stderr);
```
[the remark this answers is not in the document]
Comment (jjacob512) [reply cli/hse_cli.c:1507]: This a valid suggestion, but IMO should be delegated to a refactor of the whole cli as almost all commands follow this pattern. Will open a ticket for this.

**Comment by jjacob512** [comment]
> "this build" maybe should change to "same build"

Changed to "the same build"

**Review by beckerg — APPROVED** [review]

**Pull request merged**